### PR TITLE
Add libovsdb metrics

### PR DESCRIPTION
**Pull request by martinkennelly**
depends on ovn-org/libovsdb/pull/328

See docs/metrics.md for metric names.

I want libovsdb monitor connection metrics to create an alert if we aren't connected to SB and NB DBs.

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

Also,
Go mod vendor for vishvananda/netlink has custom modifications. When I perform go mod vendor, I need to remove the overwrites of the custom mods. It seems surya made the mods.

~~Also, fedora dockerfile is rejecting updates because of gpg check. Needed to add `--nogpgcheck` to install the packages locally.~~
Resolved - very old f36 image locally.

Testing with CI.

/cc @npinaeva

### CODE REVIEW

**Comment by martinkennelly** [comment]
Waiting for PR https://github.com/ovn-org/libovsdb/pull/329 to merge to ensure we dont register metrics more than once.

**Comment by martinkennelly** [comment]
@dcbw since you have the context on this, can you take a look?

@npinaeva noticed that some of the metrics dont seem to be exposed or ever set namely the metrics `..update_messages_total` and `..table_updates_total`. It seems their code paths are not executed. 

For `table_updates_total`, I see there are multiple update funcs here: https://github.com/ovn-org/libovsdb/blob/main/client/client.go#L418
Only if "update" is executed will this metric be updated. Do we have a visibility gap here for that metric?

For this PR, I only care about monitors and disconnections.

File: docs/metrics.md
Comment (npinaeva) [inline docs/metrics.md:20]: I just realized that `ovnkube_master_disconnects_total` doesn't tell you anything about what kind of disconnect it measures, maybe add libovsdb to the name?

**Review by npinaeva — COMMENTED** [review]
Metrics are exposed, and if disconnects=0 and monitors=2 are the expected values, it works

File: docs/metrics.md
Comment (martinkennelly) [reply docs/metrics.md:20]: True. The tooltip or description in OCP console will give context but the name isnt good.

Ill prefix it with "database", so itll be ovnkube_master_database_disconnects_total. 
Ill do the same to monitors.
I originally didnt want to rename the metrics in case anyone was using it but since they were broken anyway, why not :)

Comment (martinkennelly) [reply docs/metrics.md:20]: Ill see what dan says in libovsdb.

Comment (martinkennelly) [reply docs/metrics.md:20]: or maybe ovsdb - the reason I am not picking libovsdb is the default metric "namespace" is called "libovsdb" but for ovn-k we are overriding it with "ovnkube".

File: go-controller/pkg/libovsdb/libovsdb.go
Comment (martinkennelly) [reply go-controller/pkg/libovsdb/libovsdb.go]: I am not sure either. 
I think its fine to repeat ourselves here. 
Its unlikely this ever changes anyway.

**Comment by martinkennelly** [comment]
/hold
Dont see any sb db monitors.

File: docs/metrics.md
Comment (npinaeva) [reply docs/metrics.md]: I think new name makes perfect sense

File: go-controller/pkg/libovsdb/libovsdb.go
Comment (jcaamano) [inline go-controller/pkg/libovsdb/libovsdb.go:29]: How would you feel about passing variable length options directly to `newClient`? That way is is more generic. I have a draft where I am already doing so so it might make sense:

`func newClient(cfg config.OvnAuthConfig, dbModel model.ClientDBModel, stopCh <-chan struct{}, opts ...client.Option)`

File: go-controller/pkg/testing/libovsdb/libovsdb.go
Comment (jcaamano) [inline go-controller/pkg/testing/libovsdb/libovsdb.go:143]: I am wondering here, don't we need a stub for testing?

File: go-controller/pkg/libovsdb/libovsdb.go
Comment (martinkennelly) [reply go-controller/pkg/libovsdb/libovsdb.go:29]: Sounds much better. Can you link a PR here and I can follow and rebase once merged.

File: go-controller/pkg/testing/libovsdb/libovsdb.go
Comment (martinkennelly) [reply go-controller/pkg/testing/libovsdb/libovsdb.go:143]: I am not aware of any stub.
I didn't see the point of stubbing this out. Creating a new registry is cheap.
I can create one if you want me to.

Comment (jcaamano) [reply go-controller/pkg/testing/libovsdb/libovsdb.go:143]: > I am not aware of any stub. I didn't see the point of stubbing this out. Creating a new registry is cheap. I can create one if you want me to.

Nah, I was just wondering what happens in the tests if there is no prometheus

File: go-controller/pkg/libovsdb/libovsdb.go
Comment (jcaamano) [reply go-controller/pkg/libovsdb/libovsdb.go:29]: Is this one, but just a draft.
https://github.com/openshift/ovn-kubernetes/pull/1122

Your PR would surely go in first.

Comment (martinkennelly) [reply go-controller/pkg/libovsdb/libovsdb.go:29]: Cool. Ill make this change anyway.

File: go-controller/pkg/testing/libovsdb/libovsdb.go
Comment (martinkennelly) [reply go-controller/pkg/testing/libovsdb/libovsdb.go:143]: A little extra cpu / ram but no effect if theres no prometheus endpoint. It registers the metrics with the prometheus registry, and allows the metrics to be updated, but they are not served because theres no server serving the registry.

File: go-controller/pkg/libovsdb/libovsdb.go
Comment (martinkennelly) [reply go-controller/pkg/libovsdb/libovsdb.go]: Done

**Review by jcaamano — APPROVED** [review]
LGTM

I am thinking that we probably need some kind of `LibOVSDBClientConfig` struct that contains `OvnAuthConfig` (which I think might be more appropriately named `OVSDBAuthConfig`) and everything else we want to put in there, like the prometheus registry.

But of course I am not going to need this done in this PR, a single extra argument is not to bad for now.

But I also won't hold you from doing it if you feel like it. So giving some time.

**Comment by jcaamano** [comment]
> LGTM
> 
> I am thinking that we probably need some kind of `LibOVSDBClientConfig` struct that contains `OvnAuthConfig` (which I think might be more appropriately named `OVSDBAuthConfig`) and everything else we want to put in there, like the prometheus registry.
> 
> But of course I am not going to need this done in this PR, a single extra argument is not too bad for now.
> 
> But I also won't hold you from doing it if you feel like it. So giving some time.

@martinkennelly let me know if you are ok with merging this. Don't wanna step on any follow up work you might be doing.

**Review by trozet — APPROVED** [review]

**Comment by martinkennelly** [comment]
> LGTM
> 
> I am thinking that we probably need some kind of `LibOVSDBClientConfig` struct that contains `OvnAuthConfig` (which I think might be more appropriately named `OVSDBAuthConfig`) and everything else we want to put in there, like the prometheus registry.
> 
> But of course I am not going to need this done in this PR, a single extra argument is not too bad for now.
> 
> But I also won't hold you from doing it if you feel like it. So giving some time.

@jcaamano Thanks. I want to proceed with this as is because I am blocking someone downstream.

**Pull request merged**
